flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java

```diff
@@ -0,0 +1,224 @@
+     * @param shouldStartBufferFlushChecker indicate whether to start the buffer flushing checker
+     *     thread
+     */
+    public TieredStorageMemoryManagerImpl(
+            float numBuffersTriggerFlushRatio, boolean shouldStartBufferFlushChecker) {
+        this.numBuffersTriggerFlushRatio = numBuffersTriggerFlushRatio;
+        this.shouldStartBufferFlushChecker = shouldStartBufferFlushChecker;
+        this.tieredMemorySpecs = new HashMap<>();
+        this.numRequestedBuffers = new AtomicInteger(0);
+        this.bufferFlushCallbacks = new ArrayList<>();
+        this.isRunning = false;
+    }
+
+    @Override
+    public void setup(BufferPool bufferPool, List<TieredStorageMemorySpec> storageMemorySpecs) {
+        this.bufferPool = bufferPool;
+        for (TieredStorageMemorySpec memorySpec : storageMemorySpecs) {
+            checkState(
+                    !tieredMemorySpecs.containsKey(memorySpec.getOwner()),
+                    "Duplicated memory spec.");
+            tieredMemorySpecs.put(memorySpec.getOwner(), memorySpec);
+            numTotalGuaranteedBuffers += memorySpec.getNumGuaranteedBuffers();
+        }
+
+        if (shouldStartBufferFlushChecker) {
+            this.executor =
+                    Executors.newSingleThreadScheduledExecutor(
+                            new ThreadFactoryBuilder()
+                                    .setNameFormat("buffer flush checker")
+                                    .setUncaughtExceptionHandler(FatalExitExceptionHandler.INSTANCE)
+                                    .build());
+            this.executor.scheduleWithFixedDelay(
+                    this::checkShouldFlushCachedBuffers,
+                    DEFAULT_CHECK_FLUSH_INITIAL_DELAY_MS,
+                    DEFAULT_CHECK_FLUSH_PERIOD_DURATION_MS,
+                    TimeUnit.MILLISECONDS);
+        }
+
+        this.isRunning = true;
+    }
+
+    @Override
+    public void registerBufferFlushCallback(Runnable userBufferFlushCallBack) {
+        bufferFlushCallbacks.add(userBufferFlushCallBack);
+    }
+
+    /**
+     * Request a {@link BufferBuilder} instance from {@link BufferPool} for a specific owner. The
+     * {@link TieredStorageMemoryManagerImpl} will not check whether a buffer can be requested and
+     * only record the total number of requested buffers. If the buffers in the {@link BufferPool}
+     * is not enough, this will trigger each tiered storage to flush buffers as much as possible.
+     *
+     * @return the requested buffer
+     */
+    @Override
+    public BufferBuilder requestBufferBlocking() {
+        checkIsRunning();
+
+        MemorySegment requestedBuffer = null;
+        try {
+            requestedBuffer = bufferPool.requestMemorySegmentBlocking();
+        } catch (Throwable throwable) {
+            ExceptionUtils.rethrow(throwable, "Failed to request memory segments.");
+        }
+        numRequestedBuffers.incrementAndGet();
+        checkShouldFlushCachedBuffers();
```

Comment:
Shouldn't this happen before requesting?

Reply:
Move this before requesting. But that may not make sure the buffer reclaim occurs, so we need a checking thread.

Note that it is **not** possible to remove this, as doing so could result in the task becoming stuck in the buffer request. As the number of buffers in the buffer pool can vary at any given time, the stuck may occur if the thread is removed. 

For instance, if the memory usage of the {@link BufferPool} has been checked and {@link TieredStorageMemoryManagerImpl} determined that buffer reclamation is unnecessary, but then the buffer pool size is suddenly reduced to a very small size, the buffer request will become stuck and the task will never be able to call for buffer reclamation if this thread is removed, then a task stuck occurs.

Java doc also added these descriptions.